crates/sim2h/src/lib.rs

```diff
@@ -260,49 +231,565 @@ impl Sim2hHandle {
         &self.dht_algorithm
     }
 
-    /// acquire a mutex lock to our state data
-    pub async fn lock_state(&self) -> tokio::sync::MutexGuard<'_, Sim2hState> {
-        self.state.lock().await
+    /// access our connection manager handle
+    pub fn connection_mgr(&self) -> &ConnectionMgrHandle {
+        &self.connection_mgr
+    }
+
+    /// send a message to another connected agent
+    pub fn send(&self, agent: AgentId, uri: Lib3hUri, msg: &WireMessage) {
+        debug!(">>OUT>> {} to {}", msg.message_type(), uri);
+        MESSAGE_LOGGER
+            .lock()
+            .log_out(agent, uri.clone(), msg.clone());
+        let payload: Opaque = msg.clone().into();
+        self.connection_mgr
+            .send_data(uri, payload.as_bytes().into());
+    }
+
+    /// get access to our im_state object
+    pub fn im_state(&self) -> &sim2h_im_state::StoreHandle {
+        &self.im_state
     }
 
     /// forward a message to be handled
     pub fn handle_message(&self, uri: Lib3hUri, message: WireMessage, signer: AgentId) {
-        if let Err(e) =
-            self.send_com
-                .send(Sim2hCom::HandleMessage(Box::new(Sim2hComHandleMessage {
-                    uri,
-                    message,
-                    signer,
-                })))
-        {
-            error!("error sending message to sim2h - shutting down? {:?}", e);
-        }
-    }
+        // dispatch to correct handler
+        let sim2h_handle = self.clone();
 
-    /// forward a message to an already joined connection to be handled
-    pub fn handle_joined(
-        &self,
-        uri: Lib3hUri,
-        space_address: SpaceHash,
-        agent_id: AgentId,
-        message: WireMessage,
-    ) {
-        self.send_com
-            .send(Sim2hCom::HandleJoined(Box::new(Sim2hComHandleJoined {
-                uri,
-                space_address,
-                agent_id,
-                message,
-            })))
-            .expect("can send");
+        // these message types are allowed before joining
+        let message = match message {
+            WireMessage::Lib3hToClient(_) | WireMessage::ClientToLib3hResponse(_) =>
+                panic!("This is soo wrong. Clients should never send a message that only servers can send."),
```

Comment:
I think we should just log this as an error rather than panicing.  I don't want us to create a way that people can bring down a sim2h server, or, more likely, that we bring one down by miscoding something in the sim2h_worker side.